### PR TITLE
8304292: Memory leak related to ClassLoader::update_class_path_entry_list

### PR DESCRIPTION
**Pull request by calvinccheung**
Please review this small fix for `ClassPathEntry` leak when there's a duplicate entry in the app class path (`-cp`) and the `check_for_duplicate` argument is set to `true`, the `ClassLoader::add_to_app_classpath_entries` function should delete the entry before returning. This fix also corrects the call to `update_class_path_entry_list` from `setup_app_search_path`; the `check_for_duplicate` argument should be set to `true`. Two test cases are included to exercise the code path.

Passed tiers 1 - 4 testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304292](https://bugs.openjdk.org/browse/JDK-8304292): Memory leak related to ClassLoader::update_class_path_entry_list (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [86de8828](https://git.openjdk.org/jdk/pull/15132/files/86de882869a9a5437a3ce71d634b2b904d047f07)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15132/head:pull/15132` \
`$ git checkout pull/15132`

Update a local copy of the PR: \
`$ git checkout pull/15132` \
`$ git pull https://git.openjdk.org/jdk.git pull/15132/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15132`

View PR using the GUI difftool: \
`$ git pr show -t 15132`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15132.diff">https://git.openjdk.org/jdk/pull/15132.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15132#issuecomment-1663355130)

### CODE REVIEW

File: src/hotspot/share/classfile/classLoader.cpp [hunk 1]
```diff
@@ -521,7 +521,7 @@ void ClassLoader::setup_app_search_path(JavaThread* current, const char *class_p
 
   while (cp_stream.has_next()) {
     const char* path = cp_stream.get_next();
-    update_class_path_entry_list(current, path, false, false, false);
+    update_class_path_entry_list(current, path, true, false, false);
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/classLoader.cpp:524]: Some comments on these boolean parameters would be useful.

**Review by dholmes-ora — APPROVED** [review]
This seems fine to me. Thanks.

File: src/hotspot/share/classfile/classLoader.cpp (hunk 1 above)
Comment (calvinccheung) [reply src/hotspot/share/classfile/classLoader.cpp]: I added some comments to the above and another callsite in the same file.

**Review by dholmes-ora — APPROVED** [review]
Thanks for the comments.

File: src/hotspot/share/classfile/classLoader.cpp [hunk 2]
```diff
@@ -811,6 +813,7 @@ void ClassLoader::add_to_app_classpath_entries(JavaThread* current,
     while (e != nullptr) {
       if (strcmp(e->name(), entry->name()) == 0) {
         // entry already exists
+        delete entry;
```
Comment (iklam) [inline src/hotspot/share/classfile/classLoader.cpp:816]: I think the API contract between `update_class_path_entry_list()` and `add_to_app_classpath_entries()` is too complicated:

- `add_to_app_classpath_entries()` will delete the entry if it decides that the entry shouldn't be saved

It's better for the lifetime of the entry to be maintained by `update_class_path_entry_list()`. It should do something like:

```
 new_entry = create_class_path_entry(current, path, &st, is_boot_append, from_class_path_attr);

 if (add_to_app_classpath_entries(current, new_entry, check_for_duplicates) == false) {
  // new_entry is not saved, free it now
  delete new_entry;
 }
```

Comment (calvinccheung) [reply src/hotspot/share/classfile/classLoader.cpp]: Thanks for the review. I've pushed another commit based on your suggestions.

**Review by dholmes-ora — APPROVED** [review]
Still fine to me.

**Review by iklam — APPROVED** [review]
LGTM

**Comment by calvinccheung** [comment]
Thanks @dholmes-ora and @iklam.

/integrate

**Pull request closed without merging**